File: 461_hamming_distance/hamming_distance.go
/*
 * @Date: 2021-05-27 09:41:14
 * @Author: mengsenwang
 * @LastEditors: mengsenwang
 * @LastEditTime: 2021-05-27 09:52:36
 */

package main

import "math/bits"

func hammingDistance(x, y int) int {
	return bits.OnesCount(uint(x ^ y))
}

func hammingDistance_move_bits(x, y int) (ans int) {
	for s := x ^ y; s > 0; s >>= 1 {
		ans += s & 1
	}
	return
}

func hammingDistance_BK(x, y int) (ans int) {
	for s := x ^ y; s > 0; s &= s - 1 {
		ans++
	}
	return
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed!")
		}
	}
	assert(hammingDistance(1, 4) == 2)
	assert(hammingDistance_move_bits(1, 4) == 2)
	assert(hammingDistance_BK(1, 4) == 2)
}

File: 461_hamming_distance/hamming_distance.cpp
/*
 * @Date: 2021-05-27 09:40:52
 * @Author: mengsenwang
 * @LastEditors: mengsenwang
 * @LastEditTime: 2021-05-27 09:50:33
 */

#include <cassert>

int hammingDistance(int x, int y) { return __builtin_popcount(x ^ y); }

int hammingDistance_move_bits(int x, int y) {
  int s = x ^ y, ret = 0;
  while (s) {
    ret += s & 1;
    s >>= 1;
  }
  return ret;
}

int hammingDistance_BK(int x, int y) {
  int s = x ^ y, ret = 0;
  while (s) {
    s &= s - 1;
    ret++;
  }
  return ret;
}

int main() {
  assert(hammingDistance(1, 4) == 2);
  assert(hammingDistance_move_bits(1, 4) == 2);
  assert(hammingDistance_BK(1, 4) == 2);
  return 0;
}